### PR TITLE
Feat(PDC) Add flow cell query sorted by date

### PR DESCRIPTION
**Pull request by thelander**
### Added

- Functionality to query PDC for flow cells sorted by date and returning the latest flow cell given an ID
  - Fixes the last part of this issue https://github.com/Clinical-Genomics/cg/issues/1515: 175889260-115c5f17-9ade-48e3-9514-279f31cb2239

### How to prepare for test
- [ ] ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] paxa the environment: `paxa`
- [ ] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh feat/query-pdc-for-flow-cell-by-date`

### How to test
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Review by karlnyr — COMMENTED** [review]
does not look like this could be anything but a homerun, now I guess some practical tests and tests would fix things. I will try to see if I have a test case where there are more than 1 entry in the archive

**Comment by thelander** [comment]
@karlnyr I'll add the logging we talked about as well. What kind of information is relevant to log when retrieving a list of flow cells?

**Comment by karlnyr** [comment]
I would assume that the following could be of use IF SWEDAC would investigate the fetching of flowcells:

- Versions:
    - cg
    - dsmc version?
- Any error messages from the query, i.e. attempted fetch returning an empty search
- General command which we fetch the data from pdc with (the construct we build in cg should suffice)

@moahaegglund Would you agree?

File: cg/meta/backup/backup.py [hunk 1]
```diff
@@ -249,6 +249,17 @@ def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
         query: list = self.pdc.process.stdout.split(NEW_LINE)
         return query
 
+    def query_pdc_for_flow_cell_by_date(self, flow_cell_id) -> list:
+        """Query PDC for a given flow cell id sorted by date"""
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:253]:
```suggestion
        """Query PDC for a given flow cell id sorted by date."""
```

File: cg/meta/backup/backup.py [hunk 2]
```diff
@@ -249,6 +249,17 @@ def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
         query: list = self.pdc.process.stdout.split(NEW_LINE)
         return query
 
+    def query_pdc_for_flow_cell_by_date(self, flow_cell_id) -> list:
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:252]:
```suggestion
    def query_pdc_for_flow_cell_by_date(self, flow_cell_id) -> List[str]:
```

File: cg/meta/backup/backup.py [hunk 3]
```diff
@@ -249,6 +249,17 @@ def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
...
+        """Query PDC for a given flow cell id sorted by date"""
+        query: list = self.query_pdc_for_flow_cell(flow_cell_id)
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:254]:
```suggestion
        query: List[str] = self.query_pdc_for_flow_cell(flow_cell_id)
```

File: cg/meta/backup/backup.py [hunk 4]
```diff
@@ -249,6 +249,17 @@ def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
...
+        query.sort(key=lambda flow_cell: flow_cell.sequenced_at)
+        return query
+
+    def query_pdc_for_latest_flow_cell(self, flow_cell_id) -> str:
+        """Query PDC for the latest flow cell id"""
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:259]:
```suggestion
        """Query PDC for the latest flow cell id."""
```

File: cg/meta/backup/backup.py [hunk 5]
```diff
@@ -249,6 +249,17 @@ def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
...
+        query: list = self.query_pdc_for_flow_cell_by_date(flow_cell_id)
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:260]:
```suggestion
        query: List[str] = self.query_pdc_for_flow_cell_by_date(flow_cell_id)
```

File: cg/meta/backup/backup.py [hunk 6]
```diff
@@ -249,6 +249,17 @@ def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
...
+        return query[-1]
+
     def retrieve_archived_file(self, archived_file: Path, run_dir: Path) -> None:
         """Retrieve the archived file from PDC to a flow cell runs directory"""
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:264]:
```suggestion
        """Retrieve the archived file from PDC to a flow cell runs directory."""
```

**Review by henrikstranneheim — APPROVED** [review]
Nice!

**Comment by thelander** [comment]
> I would assume that the following could be of use IF SWEDAC would investigate the fetching of flowcells:
> 
> * Versions:
>   
>   * cg
>   * dsmc version?
> * Any error messages from the query, i.e. attempted fetch returning an empty search
> * General command which we fetch the data from pdc with (the construct we build in cg should suffice)
> 
> @moahaegglund Would you agree?

Maybe such logging is better suited for a separate PR after all?

**Comment by thelander** [comment]
@karlnyr did you find a test case with more than one entry in the archive? 😊

**Comment by karlnyr** [comment]
Hi, well of course! `210519_A00689_0271_AHVFVGDSXY` or `HVFVGDSXY` is one of the duplicated entries we have on PDC, information taken from IT-issues#305

**Comment by thelander** [comment]
> Hi, well of course! `210519_A00689_0271_AHVFVGDSXY` or `HVFVGDSXY` is one of the duplicated entries we have on PDC, information taken from IT-issues#305

@karlnyr thanks 😊

**Comment by Vince-janv** [comment]
Closing as stale

**Pull request closed without merging**